File: code/ex820.java
class Solution {
    public int minimumLengthEncoding(String[] words) {
        Arrays.sort(words, new Comparator<String>() {
            public int compare(String a, String b) {
                return b.length() - a.length();
            }
        });
        
        // Insert to trie
        int cnt = 0;
        TrieNode tn = new TrieNode();
        for (String word : words) {
            TrieNode p = tn;
            for (int i = word.length() - 1; i >= 0; i--) {
                char c = word.charAt(i);
                if (i == 0 && (p.children[c - 'a'] == null))    cnt += word.length() + 1;
                if (p.children[c - 'a'] == null) {
                    p.children[c - 'a'] = new TrieNode();
                } 
                p = p.children[c - 'a'];
            }
        }
        
        return cnt;
    }
}

class TrieNode {
    TrieNode[] children;
    TrieNode() {
        children = new TrieNode[26];
    }
}

File: code/ex820.cpp
struct TrieNode {
    TrieNode* children[26] = {nullptr};
};

struct cmp {
    bool operator()(string s1, string s2) {
        return s2.length() < s1.length();
    }
};

class Solution {
public:
    int minimumLengthEncoding(vector<string>& words) {
        sort(words.begin(), words.end(), cmp());
        TrieNode* root = new TrieNode();
        int len = 0;
        for (string word : words) {
            TrieNode* p = root;
            for (int i = word.length() - 1; i >= 0; --i) {
                char c = word[i];
                if (p->children[c - 'a'] == NULL) {
                    if (i == 0)    len += word.length() + 1;
                    p->children[c - 'a'] = new TrieNode();
                }
                p = p->children[c - 'a'];
            }
        }
        return len;
    }
};